File: Game Client Project/Assets/Script/Network/InputState.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class InputState : IMemoryStream
{
    float horizontal;
    float vertical;
    float yRotation;
    bool isShooting;
    bool isSwitchColor;
    byte mColor;

    public InputState()
    {
        this.horizontal = 0;
        this.vertical = 0;
        this.yRotation = 0;
        this.isShooting = false;
        this.isSwitchColor = false;
        this.mColor = 0;
    }

    public InputState(InputState state)
    {
        this.horizontal = state.horizontal;
        this.vertical = state.vertical;
        this.yRotation = state.yRotation;
        this.isShooting = state.isShooting;
        this.isSwitchColor = state.isSwitchColor;
        this.mColor = state.mColor;
    }

    public void Read(InputMemoryBitStream inInputStream)
    {
        inInputStream.Read(out horizontal);
        inInputStream.Read(out vertical);
        inInputStream.Read(out yRotation);
        inInputStream.Read(out isShooting);
        inInputStream.Read(out isSwitchColor);
        if (isSwitchColor)
        {
            inInputStream.Read(out mColor);
        }
        
    }

    public void Write(OutputMemoryBitStream inOutputStream)
    {
        inOutputStream.Write(horizontal);
        inOutputStream.Write(vertical);
        inOutputStream.Write(yRotation);
        inOutputStream.Write(isShooting);
        inOutputStream.Write(isSwitchColor);
        if (isSwitchColor)
        {
            inOutputStream.Write(mColor);
        }
    }

    public void UpdateInputValue(float horizontal, float vertical, float yRotation, bool isShooting, bool isSwitchColor, byte mColor)
    {
        this.horizontal = horizontal;
        this.vertical = vertical;
        this.yRotation = yRotation;
        this.isShooting = isShooting;
        this.isSwitchColor = isSwitchColor;
        this.mColor = mColor;
    }

    void WriteSignedBinaryValue(OutputMemoryBitStream inOutputStream, float inValue)
    {
        bool isNonZero = (inValue != 0f);
        inOutputStream.Write(isNonZero);
        if (isNonZero)
        {
            inOutputStream.Write(inValue > 0f);
        }
    }

    void ReadSignedBinaryValue(InputMemoryBitStream inInputStream, out float outValue )
    {
        bool isNonZero;
        inInputStream.Read(out isNonZero);
        if (isNonZero)
        {
            bool isPositive;
            inInputStream.Read(out isPositive);
            outValue = isPositive ? 1f : -1f;
        }
        else
        {
            outValue = 0f;
        }
    }
}

File: GameServer/RoboCat/Src/InputState.cpp
#include <RoboCatPCH.h>

namespace
{
	void WriteSignedBinaryValue( OutputMemoryBitStream& inOutputStream, float inValue )
	{
		bool isNonZero = ( inValue != 0.f );
		inOutputStream.Write( isNonZero );
		if( isNonZero )
		{
			inOutputStream.Write( inValue > 0.f );
		}
	}

	void ReadSignedBinaryValue( InputMemoryBitStream& inInputStream, float& outValue )
	{
		bool isNonZero;
		inInputStream.Read( isNonZero );
		if( isNonZero )
		{
			bool isPositive;
			inInputStream.Read( isPositive );
			outValue = isPositive ? 1.f : -1.f;
		}
		else
		{
			outValue = 0.f;
		}
	}
}

bool InputState::Write( OutputMemoryBitStream& inOutputStream ) const
{
	//WriteSignedBinaryValue( inOutputStream, GetDesiredHorizontalDelta() );
	//WriteSignedBinaryValue( inOutputStream, GetDesiredVerticalDelta() );
	inOutputStream.Write( GetDesiredHorizontalDelta() );
	inOutputStream.Write( GetDesiredVerticalDelta() );
	inOutputStream.Write( mRotation );
	inOutputStream.Write( mIsShooting );
	inOutputStream.Write( mIsSwitchColor );
	if (mIsSwitchColor)
	{
		inOutputStream.Write(mColor);
	}

	return false;
}

bool InputState::Read( InputMemoryBitStream& inInputStream )
{
	//ReadSignedBinaryValue( inInputStream, mDesiredRightAmount );
	//ReadSignedBinaryValue( inInputStream, mDesiredForwardAmount );
	inInputStream.Read( mHorizontal );
	inInputStream.Read( mVertical );
	inInputStream.Read( mRotation );
	inInputStream.Read( mIsShooting );
	inInputStream.Read( mIsSwitchColor );
	if (mIsSwitchColor)
	{
		inInputStream.Read(mColor);
	}

	return true;
}